Arch arm core assembly fixes

ioannisg opened the pull request:
Several fixes in ARM Core assembly, mostly documentation/performance. 
One fix in 5f4d057 is non-trivial (but not critical), therefore, registering this as bug-fix with priority medium.

ioannisg committed:
arch: arm: remove unnecessary stacking of r3 in swap.S

Stacking r3 at this point is not required as the register value
is not used afterwards, therefore, it does not need to be
retained.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: remove unnecessary stacking/unstacking

Due to code re-organization inside __pendSV(), the stacking
and unstacking of scratch registers around the invocation of
read_timer_end_of_swap() is no longer required.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
kernel: minor syntax fix in Kconfig

Minor style (syntax) fix in the help text of symbol
config EXECUTION_BENCHMARKING.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: update function documentation for __svc

Fixing the documentation for ARM core __svc function, which was
outdated and present only for the ARMv7-M version.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: ARMv8-M built-in stack protection fixes in userspace.S

This commit provides two fixes for the ARMv8-M built-in stack
protection in userspace.S. First, it clears the PSPLIM register
before setting the PSP to the privileged stack. In the wake of
this, the privilege stack area can now be positioned enywhere,
i.e. does not have to be above the user stack area. Second, it
removes unncessary push/pop when clearing the PSPLIM register,
upon returning from the system call, since the r3 register is
not used anymore in this routine.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm remove unnecessary stacking of r3 in userspace.S

This commit moves the clearing of PSPLIM before determining the
privilege stack start and size. In this way the clearing can be
done without unnecessary pushing and popping r3 core register.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

agross-oss reviewed (approved):
kernel: minor syntax fix in Kconfig

Minor style (syntax) fixe in the help text of symbol
config EXECUTION_BENCHMARKING.

Aside from a nit to fix fixe in the above commit message, it looks good.

ioannisg committed:
arch: arm: invoke MPU reprogramming while in privilege stack

_arch_configure_dynamic_mpu_regions() may re-program the MPU
Stack Guard, to guard the privilege stack for overflows,
(if building with option CONFIG_MPU_STACK_GUARD). There is a
risk of actually overflowing the stack while doing the
re-programming. We minimize the risk by placing the
re-programming function immediately after we have switched to
the privileged stack so that the whole privilege stack area is
available for this critical operation.

Note that the risk for overflow is higher if using the normal
thread stack, since we do not control how much stack is actually
left, when user invokes _arm_userspace_enter().

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
I included an additional commit in this PR that requires the attention of @agross-linaro , once more!

agross-linaro commented on arch/arm/core/userspace.S:
```diff
@@ -73,6 +64,28 @@ SECTION_FUNC(TEXT,_arm_userspace_enter)
+     * stack, since we do not control how much stack is actually left, when
+     * user invokes _arm_userspace_enter().
+     */
+    push {r1,r2,r3,ip,lr}
```
need 8 byte aligned pushes before function calls.  otherwise stack could go non 8 byte aligned.

ioannisg replied:
This is not exactly before the BL instruction though, @agross-linaro

ioannisg replied:
Fixed (Added r0, to make 8-byte aligned push)

agross-oss reviewed (changes requested):
just need to fix the number of push/pops

ioannisg commented:
@galak @MaureenHelm could we get this one in? 
Intel folks are invited to review this one, only due to minor typo fixes in kernel Kconfig. :)
